### PR TITLE
1.20.2 Support

### PR DESCRIPTION
**Pull request by Kas-tle**
Also added support for Eclipse IDE so I could actually work on it...

### Spigot

- `PacketHandshakingInSetProtocol` is now a record so we must construct call its constructor instead of modifying
  - Unfortunately this requires 2 ints with the same access level, so we must specify obfuscated field names
- `initUUID` gone and that info is now passed directly to the event via `LoginHandler().fireEvents(GameProfile)`
- There are now multiple `PacketListener` fields in `NetworkManager`, so we must specify the obfuscated field name

### Paper
- Also works with proposed Spigot changes per their experimental builds

### Velocity

Note that this is based on the current PR for 1.20.2 support on the Velocity repo: https://github.com/PaperMC/Velocity/pull/1088
- Bump API to `3.2.0-SNAPSHOT`
- Bump minimum Java version to `11` per Velocity API
- `getSessionHandler` is now `getActiveSessionHandler`

### Bungeecord

Works without changes (truly shocking).

### CODE REVIEW

File: spigot/src/main/java/org/geysermc/floodgate/addon/data/SpigotDataHandler.java [hunk 1]
```diff
@@ -58,8 +59,28 @@ protected void setNewIp(Channel channel, InetSocketAddress newIp) {
...
+                Object[] components = new Object[]{
+                    ClassNames.HANDSHAKE_PORT.get(handshakePacket),
+                    hostname,
+                    ClassNames.HANDSHAKE_PROTOCOL.get(handshakePacket),
+                    ClassNames.HANDSHAKE_INTENTION.get(handshakePacket)
+                };
+
+                return ClassNames.HANDSHAKE_PACKET_CONSTRUCTOR.newInstance(components);
+            } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+                e.printStackTrace();
```
Comment (Tim203) [inline spigot/src/main/java/org/geysermc/floodgate/addon/data/SpigotDataHandler.java:79]: I'd just rethrow the exception if something goes wrong

File: spigot/src/main/java/org/geysermc/floodgate/util/ClassNames.java [hunk 2]
```diff
@@ -93,7 +103,7 @@ public class ClassNames {
 
         String nmsPackage = SPIGOT_MAPPING_PREFIX + '.';
 
-
+
```
Comment (Tim203) [inline spigot/src/main/java/org/geysermc/floodgate/util/ClassNames.java:106]: Spaces not needed

File: spigot/src/main/java/org/geysermc/floodgate/util/ClassNames.java [hunk 3]
```diff
@@ -229,9 +252,50 @@ public class ClassNames {
...
     }
 
     private static <T> T checkNotNull(@CheckForNull T toCheck, @CheckForNull String objectName) {
         return Preconditions.checkNotNull(toCheck, objectName + " cannot be null");
     }
+
+    // Ensure one of two is not null
+    private static <T> T checkNotNull(@CheckForNull T toCheck, @CheckForNull T toCheck2,
+                                      @CheckForNull String objectName,
+                                      @CheckForNull String objectName2) {
```
Comment (Tim203) [inline spigot/src/main/java/org/geysermc/floodgate/util/ClassNames.java:297]: If the limit is reached just do one argument per line like the SpigotCommandUtil constructor for example

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -22,6 +22,27 @@ jobs:
...
+          name: Floodgate Spigot
+          path: spigot/build/libs/floodgate-spigot.jar
+  
+      - name: Archive artifacts (Floodgate Velocity)
+        uses: actions/upload-artifact@v2
+        if: success()
+        with:
+          name: Floodgate Velocity
+          path: velocity/build/libs/floodgate-velocity.jar
+
```
Comment (Tim203) [inline .github/workflows/build.yml:45]: I think that just the downloads api & maven is enough

**Review by Tim203 — CHANGES_REQUESTED** [review]
Thanks for the work, just a few nitpicks

File: .github/workflows/build.yml (hunk 4 above)
Comment (Kas-tle) [reply .github/workflows/build.yml]: https://github.com/GeyserMC/Floodgate/pull/449/commits/2c52392451dc6af607b9b2194f6f460ecaff96a8#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721

File: spigot/src/main/java/org/geysermc/floodgate/util/ClassNames.java (hunk 2 above)
Comment (Kas-tle) [reply spigot/src/main/java/org/geysermc/floodgate/util/ClassNames.java]: https://github.com/GeyserMC/Floodgate/pull/449/commits/2c52392451dc6af607b9b2194f6f460ecaff96a8#diff-c9d452f82010aef19e8cd407cb82aa21dff9a8fc01101c649ae2acb6e833b7e5R106

File: spigot/src/main/java/org/geysermc/floodgate/util/ClassNames.java (hunk 3 above)
Comment (Kas-tle) [reply spigot/src/main/java/org/geysermc/floodgate/util/ClassNames.java]: https://github.com/GeyserMC/Floodgate/pull/449/commits/2c52392451dc6af607b9b2194f6f460ecaff96a8#diff-c9d452f82010aef19e8cd407cb82aa21dff9a8fc01101c649ae2acb6e833b7e5R295-R300

File: spigot/src/main/java/org/geysermc/floodgate/addon/data/SpigotDataHandler.java (hunk 1 above)
Comment (Kas-tle) [reply spigot/src/main/java/org/geysermc/floodgate/addon/data/SpigotDataHandler.java]: https://github.com/GeyserMC/Floodgate/pull/449/commits/55c03e3804ac57b025a9cd7bfa5f6cf58746cef3#diff-a83ed678d5742d28bade114ac629507328659171b3c3f3440f22c18066b7bdbfR79

**Review by Tim203 — APPROVED** [review]

**Pull request merged**